src/lib.rs

```diff
@@ -1576,6 +1629,62 @@ mod test {
             test_assign(_1_2, 1, _3_2);
         }
 
+        #[test]
+        fn test_add_overflow() {
+            // compares Ratio(1, T::max_value()) + Ratio(1, T::max_value())
+            // to Ratio(1+1, T::max_value()) for each integer type.
+            // Previously, this calculation would overflow.
+            fn test_add_typed_overflow<T>()
+            where
+                T: Integer + Bounded + Clone + Debug + CheckedAdd + CheckedMul,
+            {
+                let _6 = T::one() + T::one() + T::one() + T::one() + T::one() + T::one();
+                let _1_max = Ratio::new(T::one(), T::max_value());
+                let _2_max = Ratio::new(T::one() + T::one(), T::max_value());
+                let _6_max = Ratio::new(_6, T::max_value());
+                assert_eq!(_1_max.clone() + _1_max.clone(), _2_max);
+                assert_eq!(_1_max.clone().checked_add(&_1_max), Some(_2_max.clone()));
+                assert_eq!(
+                    _1_max.clone()
+                        + _1_max.clone()
+                        + _1_max.clone()
+                        + _1_max.clone()
+                        + _1_max.clone()
+                        + _1_max.clone(),
+                    _6_max
+                );
+                assert_eq!(
+                    _1_max
+                        .clone()
+                        .checked_add(&_1_max)
+                        .unwrap()
+                        .checked_add(&_1_max)
+                        .unwrap()
+                        .checked_add(&_1_max)
+                        .unwrap()
+                        .checked_add(&_1_max)
+                        .unwrap()
+                        .checked_add(&_1_max),
+                    Some(_6_max.clone())
+                );
+            }
+            test_add_typed_overflow::<u8>();
+            test_add_typed_overflow::<u16>();
+            test_add_typed_overflow::<u32>();
+            test_add_typed_overflow::<u64>();
+            test_add_typed_overflow::<usize>();
+            #[cfg(has_u128)]
+            test_add_typed_overflow::<u128>();
+
+            test_add_typed_overflow::<i8>();
+            test_add_typed_overflow::<i16>();
+            test_add_typed_overflow::<i32>();
+            test_add_typed_overflow::<i64>();
+            test_add_typed_overflow::<isize>();
+            #[cfg(has_i128)]
+            test_add_typed_overflow::<i128>();
+        }
```

Comment:
What do you think of this approach to testing -- using many of Rust's built-in numeric types? It seems like most of the other tests are just being run for Rational = Ratio<isize>? I think it makes sense to run most if not all tests against a signed and unsigned Ratio type. e.g. Ratio<isize> and Ratio<usize>. I suspect some bugs may show up only on signed and others only on unsigned.